htlcswitch/circuit_map.go

```diff
@@ -250,6 +255,214 @@ func (cm *circuitMap) initBuckets() error {
+			// closed channel ID map.
+			if !isClosedChannel(circuit.Incoming.ChanID) {
+				return nil
+			}
+
+			circuitKeySet[circuit.Incoming] = struct{}{}
+
+			return nil
+		}); err != nil {
+			return err
+		}
+
+		// If a keystone's InKey or OutKey matches the short channel id
+		// in the closed channel ID map, it will be deleted.
+		err := keystoneBkt.ForEach(func(k, v []byte) error {
+			var (
+				inKey  CircuitKey
+				outKey CircuitKey
+			)
+
+			// Decode the incoming and outgoing circuit keys.
+			if err := inKey.SetBytes(v); err != nil {
+				return err
+			}
+			if err := outKey.SetBytes(k); err != nil {
+				return err
+			}
+
+			// Check if the incoming channel ID can be found in the
+			// closed channel ID map.
+			if isClosedChannel(inKey.ChanID) {
+				// If the incoming channel is closed, we can
+				// skip checking on outgoing channel ID because
+				// this keystone will be deleted.
+				keystoneKeySet[outKey] = struct{}{}
+
+				// Technically the incoming keys found in
+				// keystone bucket should be a subset of
+				// circuit bucket. So a previous loop should
+				// have this inKey put inside circuitAddKey map
+				// already. We do this again to be sure the
+				// circuits are properly cleaned. Even this
+				// inKey doesn't exist in circuit bucket, we
+				// are fine as db deletion is a noop.
+				circuitKeySet[inKey] = struct{}{}
+				return nil
+			}
+
+			// Check if the outgoing channel ID can be found in the
+			// closed channel ID map. Notice that we need to store
+			// the outgoing key because it's used for db query.
+			if isClosedChannel(outKey.ChanID) {
+				keystoneKeySet[outKey] = struct{}{}
+
+				// Also update circuitKeySet to mark the
+				// payment circuit needs to be deleted.
+				circuitKeySet[inKey] = struct{}{}
+			}
+
+			return nil
+		})
+		return err
+
+	}, func() {
+		// Reset the sets.
+		circuitKeySet = make(map[CircuitKey]struct{})
+		keystoneKeySet = make(map[CircuitKey]struct{})
+	}); err != nil {
+		return err
+	}
+
+	log.Debugf("To be deleted: num_circuits=%v, num_keystones=%v",
+		len(circuitKeySet), len(keystoneKeySet),
+	)
+
+	numCircuitsDeleted := 0
+	numKeystonesDeleted := 0
+
+	// Delete all the circuits and keystones for closed channels.
+	if err := kvdb.Update(cm.cfg.DB, func(tx kvdb.RwTx) error {
+		circuitBkt := tx.ReadWriteBucket(circuitAddKey)
+		if circuitBkt == nil {
+			return ErrCorruptedCircuitMap
+		}
+		keystoneBkt := tx.ReadWriteBucket(circuitKeystoneKey)
+		if keystoneBkt == nil {
+			return ErrCorruptedCircuitMap
+		}
+
+		// Delete the ciruit.
+		for inKey := range circuitKeySet {
+			if err := circuitBkt.Delete(inKey.Bytes()); err != nil {
+				return err
+			}
+
+			numCircuitsDeleted++
+		}
+
+		// Delete the keystone using the outgoing key.
+		for outKey := range keystoneKeySet {
+			err := keystoneBkt.Delete(outKey.Bytes())
+			if err != nil {
+				return err
+			}
+
+			numKeystonesDeleted++
+		}
+
+		return nil
+	}, func() {}); err != nil {
```

Comment:
The closure counter variables (`numKeystonesDeleted`, etc) should also be reset each time to avoid logging some possibly incorrect values.

Reply:
Done!